File: backend/app/api/v1/content.py
"""Public content API — home promo + hero (Phase J)."""

from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import db_session
from app.api.v1.schemas.content import (
    HeroSlideOut,
    HomeContentOut,
    LifestyleTileOut,
    PromoSettingsOut,
)
from app.application.content.service import get_home_content

router = APIRouter(prefix="/content", tags=["content"])


def _to_out(data: dict) -> HomeContentOut:
    settings = data.get("promoSettings") or data.get("promo_settings") or {}
    if not isinstance(settings, dict):
        settings = {}
    slides = []
    for s in data.get("heroSlides") or []:
        if not isinstance(s, dict):
            continue
        slides.append(
            HeroSlideOut(
                id=str(s.get("id") or ""),
                title_line1=str(s.get("titleLine1") or s.get("title_line1") or ""),
                title_line2=str(s.get("titleLine2") or s.get("title_line2") or ""),
                title_accent=s.get("titleAccent") or s.get("title_accent"),
                subtitle=str(s.get("subtitle") or ""),
                cta_label=str(s.get("ctaLabel") or s.get("cta_label") or ""),
                cta_href=str(s.get("ctaHref") or s.get("cta_href") or ""),
                secondary_label=s.get("secondaryLabel") or s.get("secondary_label"),
                secondary_href=s.get("secondaryHref") or s.get("secondary_href"),
                image_url=str(s.get("imageUrl") or s.get("image_url") or ""),
                image_alt=str(s.get("imageAlt") or s.get("image_alt") or ""),
            )
        )
    lifestyle = []
    for t in data.get("lifestyleCollections") or data.get("lifestyle_collections") or []:
        if not isinstance(t, dict):
            continue
        lifestyle.append(
            LifestyleTileOut(
                id=str(t.get("id") or ""),
                title=str(t.get("title") or ""),
                href=str(t.get("href") or ""),
                image_url=str(t.get("imageUrl") or t.get("image_url") or ""),
                span=str(t.get("span") or "square"),
            )
        )
    return HomeContentOut(
        promo_messages=list(data.get("promoMessages") or []),
        promo_settings=PromoSettingsOut(
            speed_seconds=int(
                settings.get("speedSeconds") or settings.get("speed_seconds") or 32
            ),
            color=str(settings.get("color") or "#3a7ca5"),
        ),
        hero_slides=slides,
        lifestyle_collections=lifestyle,
        updated_at=data.get("updatedAt"),
    )


@router.get("/home", response_model=HomeContentOut)
async def public_home_content(session: AsyncSession = Depends(db_session)) -> HomeContentOut:
    """Public storefront home content (promo ticker + hero slides)."""
    data = await get_home_content(session)
    return _to_out(data)

File: src/lib/api/content.ts
/**
 * Public home content client (Phase J).
 * Tags: `home`, `content` — revalidated after admin save.
 */

import { getApiBaseUrl } from "@/lib/api/config";
import { fallbackHomeContent } from "@/lib/content/defaults";
import type { HeroSlide, HomeContent, LifestyleTile } from "@/types/content";

const DEFAULT_REVALIDATE = 60;

function asString(v: unknown, fallback = ""): string {
  return typeof v === "string" ? v : fallback;
}

function normalizeSlide(raw: Record<string, unknown>): HeroSlide | null {
  const id = asString(raw.id);
  const titleLine1 = asString(raw.titleLine1 ?? raw.title_line1);
  const titleLine2 = asString(raw.titleLine2 ?? raw.title_line2);
  const subtitle = asString(raw.subtitle);
  const ctaLabel = asString(raw.ctaLabel ?? raw.cta_label);
  const ctaHref = asString(raw.ctaHref ?? raw.cta_href);
  const imageUrl = asString(raw.imageUrl ?? raw.image_url);
  if (!id || !titleLine1 || !titleLine2 || !ctaHref || !imageUrl) return null;

  const slide: HeroSlide = {
    id,
    titleLine1,
    titleLine2,
    subtitle,
    ctaLabel: ctaLabel || "Shop",
    ctaHref,
    imageUrl,
    imageAlt: asString(raw.imageAlt ?? raw.image_alt, titleLine1),
  };

  const accent = raw.titleAccent ?? raw.title_accent;
  if (typeof accent === "string" && accent.trim()) {
    slide.titleAccent = accent.trim();
  }
  const secLabel = raw.secondaryLabel ?? raw.secondary_label;
  const secHref = raw.secondaryHref ?? raw.secondary_href;
  if (typeof secLabel === "string" && secLabel.trim()) {
    slide.secondaryLabel = secLabel.trim();
  }
  if (typeof secHref === "string" && secHref.trim()) {
    slide.secondaryHref = secHref.trim();
  }
  return slide;
}

function normalizeLifestyle(raw: Record<string, unknown>): LifestyleTile | null {
  const id = asString(raw.id);
  const title = asString(raw.title);
  const href = asString(raw.href);
  const imageUrl = asString(raw.imageUrl ?? raw.image_url);
  if (!id || !title || !href || !imageUrl) return null;
  const spanRaw = asString(raw.span, "square").toLowerCase();
  const span =
    spanRaw === "tall" || spanRaw === "wide" || spanRaw === "square"
      ? spanRaw
      : "square";
  return { id, title, href, imageUrl, span };
}

export function normalizeHomeContent(raw: Record<string, unknown>): HomeContent {
  const promosRaw = raw.promoMessages ?? raw.promo_messages;
  const slidesRaw = raw.heroSlides ?? raw.hero_slides;
  const lifeRaw = raw.lifestyleCollections ?? raw.lifestyle_collections;

  const promoMessages = Array.isArray(promosRaw)
    ? promosRaw
        .filter((m): m is string => typeof m === "string")
        .map((m) => m.trim())
        .filter(Boolean)
    : [];

  const heroSlides: HeroSlide[] = [];
  if (Array.isArray(slidesRaw)) {
    for (const item of slidesRaw) {
      if (item && typeof item === "object") {
        const slide = normalizeSlide(item as Record<string, unknown>);
        if (slide) heroSlides.push(slide);
      }
    }
  }

  const lifestyleCollections: LifestyleTile[] = [];
  if (Array.isArray(lifeRaw)) {
    for (const item of lifeRaw) {
      if (item && typeof item === "object") {
        const tile = normalizeLifestyle(item as Record<string, unknown>);
        if (tile) lifestyleCollections.push(tile);
      }
    }
  }

  return {
    promoMessages,
    heroSlides,
    lifestyleCollections,
    updatedAt:
      raw.updatedAt == null && raw.updated_at == null
        ? null
        : asString(raw.updatedAt ?? raw.updated_at, "") || null,
  };
}

/**
 * Fetch home CMS content. On failure returns fallback (site never blank).
 * Uses Next fetch cache tags for admin revalidate.
 */
export async function getHomeContent(): Promise<HomeContent> {
  const base = getApiBaseUrl();
  try {
    const res = await fetch(`${base}/api/v1/content/home`, {
      headers: { Accept: "application/json" },
      next: {
        revalidate: DEFAULT_REVALIDATE,
        tags: ["home", "content"],
      },
    });
    if (!res.ok) {
      console.error("[content] home fetch failed", res.status);
      return fallbackHomeContent();
    }
    const data = (await res.json()) as Record<string, unknown>;
    // Empty promo/hero is a valid clean storefront — do not re-inject demo fallbacks.
    return normalizeHomeContent(data);
  } catch (e) {
    console.error("[content] home fetch error", e);
    return fallbackHomeContent();
  }
}
